test/e2e/framework/framework.go

```diff
@@ -0,0 +1,210 @@
+		ClientSet: client,
+	}
+
+	BeforeEach(f.BeforeEach)
+	AfterEach(f.AfterEach)
+
+	return f
+}
+
+// CreateNamespace creates a e2e test namespace.
+func (f *Framework) CreateNamespace(baseName string, labels map[string]string) (*v1.Namespace, error) {
+	if labels == nil {
+		labels = map[string]string{}
+	}
+
+	namespaceObj := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("ccm-e2e-tests-%v-", baseName),
+			Namespace:    "",
+			Labels:       labels,
+		},
+		Status: v1.NamespaceStatus{},
+	}
+
+	// Be robust about making the namespace creation call.
+	var got *v1.Namespace
+	if err := wait.PollImmediate(Poll, 30*time.Second, func() (bool, error) {
+		var err error
+		got, err = f.ClientSet.CoreV1().Namespaces().Create(namespaceObj)
+		if err != nil {
+			Logf("Unexpected error while creating namespace: %v", err)
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		return nil, err
+	}
+
+	if got != nil {
+		f.namespacesToDelete = append(f.namespacesToDelete, got)
+	}
+
+	return got, nil
+}
+
+// DeleteNamespace deletes a given namespace and waits until its contents are
+// deleted.
+func (f *Framework) DeleteNamespace(namespace string, timeout time.Duration) error {
+	startTime := time.Now()
+	if err := f.ClientSet.CoreV1().Namespaces().Delete(namespace, nil); err != nil {
+		return err
+	}
+
+	// wait for namespace to delete or timeout.
+	err := wait.PollImmediate(Poll, timeout, func() (bool, error) {
+		if _, err := f.ClientSet.CoreV1().Namespaces().Get(namespace, metav1.GetOptions{}); err != nil {
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			Logf("Error while waiting for namespace to be terminated: %v", err)
+			return false, nil
+		}
+		return false, nil
+	})
+
+	// Namespace deletion timed out.
+	if err != nil {
+		return fmt.Errorf("namespace %v was not deleted with limit: %v", namespace, err)
+	}
+
+	Logf("namespace %v deletion completed in %s", namespace, time.Now().Sub(startTime))
+	return nil
+}
+
+// BeforeEach gets a client and makes a namespace.
+func (f *Framework) BeforeEach() {
+	// The fact that we need this feels like a bug in ginkgo.
+	// https://github.com/onsi/ginkgo/issues/222
+	f.cleanupHandle = AddCleanupAction(f.AfterEach)
+	if f.ClientSet == nil {
+		By("Creating a kubernetes client")
+		config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+		Expect(err).NotTo(HaveOccurred())
+		f.ClientSet, err = clientset.NewForConfig(config)
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	By("Building a namespace api object")
+	namespace, err := f.CreateNamespace(f.BaseName, map[string]string{
+		"e2e-framework": f.BaseName,
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	f.Namespace = namespace
+}
+
+// AfterEach deletes the namespace(s).
+func (f *Framework) AfterEach() {
+	RemoveCleanupAction(f.cleanupHandle)
+
+	// DeleteNamespace at the very end in defer, to avoid any expectation
+	// failures preventing deleting the namespace.
+	defer func() {
```

Comment:
Why is this in a defer? Nothing else is being done.

Reply:
Good point! I pulled it over piecemeal from core but removed its reason for existence 🤦‍♂️